### DISCUSSION TITLE
Generic I2C driver

### DISCUSSION THREAD

**Discussion by HouseOnTheHill** [arendst/Tasmota#11973] [Q&A] [answered]
Hi all

**Background**
I built a generic PCB with an ESP-07 that brings out all the IO (5V) and supports a daughter board via I2C. The daughter board has one PCF8574 and two ADS1115 and one PCF8583 set up as an event counter with a supercap.

The PCF8583 is actually a RTC, but it also has this cool feature where it can count events like a water meter.

**Question**
Tasmota does not have a driver for the PCF8583, and I am a bit intimidated by the prospect of writing one. Is there some way of sending I2C commands to the PCF8583 and reading registers?

That might be an easier way for me to achieve the goal.

Thanks
Pelican

**Comment by sfromis** [comment, marked as answer]
With ESP32, Tasmota has the Berry scripting language including support for creating I2C drivers (I've done that), [example](https://tasmota.github.io/docs/Berry-Cookbook/#creating-a-tasmota-driver). Berry is a much more forgiving environment than creating a driver in C++. I'm not likely to start doing C++ drivers, but if I got that idea, i'd still be inclined to prototype it in Berry first, to get the basic protocol to work.

Are you aware that Tasmota can count events (like water meter pulses) on its own?

**Reply by HouseOnTheHill** [reply]
Thanks for the idea. 
I have already gone down the road of using the ESP-07S, so the ESP32 won't help me.
The ESP also can't count in power down. The PCF8583 can run off a supercap for ages.

**Reply by gemu2015** [reply]
tasmota esp8266 scripting also supports i2c. it supports up to 6.2 kb script on 1M devices

an example of reading an AXP power chip
\>D
volt=0
curr=0
found=0
\>B
; check device on I2C bus Nr.1
found=ia(0x34)
\>S
; if found read registers, (this example takes 2 ms to read both values)
if found>0 {
volt=ir(0x5a)<<4|ir(0x5b)*1.7/1000
curr=ir(0x58)<<4|ir(0x59)*0.625
}
\>W
; show on webui
Bus Voltage{m}%volt% V
Bus Current{m}%curr% mA

**Reply by gemu2015** [reply]
looking at the data sheet the counter mode is extremely easy.
set mode to 0x20 and read 6 decimal nibbles.

\>D
counter=0
found=0
\>B
; check device on I2C, Address depends
found=ia(0xA0)
if found>0 {
; set to counter mode 0 = mode register
iw(0 0x20)
}
\>S
; if found read registers
if found>0 {
; read 3 bytes from location 1 into 24 bit word
counter=ir3(1)
; get the 6 decimal nibbles
counter=(counter>>20&0xf)+(counter>>16&0xf\*10)+(counter>>12&0xf\*100)+(counter>>8&0xf\*1000)+(counter>>4&0xf\*10000)+(counter&0xf\*100000)
}
\>W
; show on webui
counter{m}%counter%

**Reply by HouseOnTheHill** [reply]
Wow - thank you for this. I will definitely try it.

**Reply by gemu2015** [reply]
much shorter converter for nibbles:

raw=ir3(1)
counter=0
fac=1
for cnt 20 0 -4
counter+=raw>>cnt&0xf\*fac
fac\*=10
next

**Reply by HouseOnTheHill** [reply]
Hey @gemu2015 
I built Tasmota with the scripting enabled and have been playing with it. 

I am getting an error:
08:29:03.872 Script: nv=2, tv=0, vns=14, ram=52
08:29:03.874 var not found: ia(0xA0)

So this means that Tasmota doesn't see the PCF8583 counter. I also tried at 0xA2 with the same result. So then I tried to see if I can see up the devices at 0x20 (PCF8574) and 0x48 and 0x49 (2 x ADS1115) but I get the same result. Tasmota picks the PCF8574 and ADS1115's up, so the bus is OK and I suspect the issue is probably related to the build.

**Reply by gemu2015** [reply]
you must use a compiler switch to enable i2c support in scripting

#define USE_SCRIPT_I2C

**Reply by HouseOnTheHill** [reply]
Sorry - I didn't add the switches - I currently have all three I need (that I know of):

#ifndef USE_SCRIPT
#define USE_SCRIPT  
#endif

#ifndef USE_SCRIPT_I2C
#define USE_SCRIPT_I2C  
#endif

#ifndef USE_SCRIPT_WEB_DISPLAY
#define USE_SCRIPT_WEB_DISPLAY
#endif

**Reply by sfromis** [reply]
IIRC, you also need to be on a rather recent level. Did you try with a fresh development source pull?

**Reply by HouseOnTheHill** [reply]
Yup - I am on 9.4.0
How do I confirm that USE_SCRIPT_I2C was in fact defined during the build?
I know USE_SCRIPT was because I have the feature and I know USE_SCRIPT_WEB_DISPLAY was because I can display something.
I have been tracing back on the board and confirmed that the I2C sensors work 100% and that PCF is hooked up correctly.

What is really puzzling is that I can't read **any** I2C address (even from the working devices) using the scripting.

Is it possible that you can't use I2C scripting **and** I2C sensors at the same time?

**Reply by gemu2015** [reply]
i just compiled on an ESP8266 and it works without problems.
do you still get var not found ?
that means it is not compiled ?
use the VERY last DEV version. i2c support is online since a few days only

**Reply by HouseOnTheHill** [reply]
Wow - thank you for the effort.
I still get the same error. I also tried without the sensors but it is the same.

Let me try with the latest dev version...

**Reply by HouseOnTheHill** [reply]
BTW I think I will have to change the address to 0xA1 first to read the registers, but let me get the right dev version going first.

**Reply by barbudor** [reply]
It is quite common to consider the i2c address without the read/write bit. So you may have to use 0x50

**Reply by barbudor** [reply]
An i2cscan in the console should confirm the address

**Reply by HouseOnTheHill** [reply]
Progress! I grabbed 9.4.0.3 and I don't get the error anymore.
I get a device on 0xA1, but not on 0xA0. This is strange as these are the two addresses for read and write.

**Reply by HouseOnTheHill** [reply]
Ah OK - saw your comment on the read/write bit. i2cscan picks up the device without this bit at 0x50. Trying that now

**Reply by HouseOnTheHill** [reply]
It works!
Thank you @barbudor @gemu2015  and @sfromis

**Reply by HouseOnTheHill** [reply]
Some feedback on my experience with this:

1) The counter nibbles are swapped, so you have to change this line to:
counter=(counter>>20&0xf*10)+(counter>>16&0xf*1)+(counter>>12&0xf*1000)+(counter>>8&0xf*100)+(counter>>4&0xf*100000)+(counter&0xf*10000)

2) You have to add the following to add the value into the JSON sent on MQTT teleperiod:
>J
; send to mqtt
,"counter":%0counter%

3) You have to set up a manual MQTT sensor in your configuration.yaml as there is no way to add this value pair to the auto discovery. I am guessing there is a manual way to set up a sensor using autodiscovery with script, but I do not know how to do this yet.

4) My board also has a PCF8574 io expander and a ADS1115 device. The Tasmota drivers append stuff after the device number, so I get "ADS1115-48" and "PCF8574-1" in the JSON (returned via MQTT). The dash ("-") messes with the templating. If anybody knows how to deal with this without going to Node Red first, then please let me know.

5) The ADS1115 driver works with the autodiscovery, but not the PCF8574 driver so you have to set up the sensors in configuration.yaml. The most elegant way of solving the discovery of the PCF8574 and PCF8583, might be to find a way to send the teleperiod transmission as an autodiscovery on boot.

Thanks again to everybody who helped.

**Reply by barbudor** [reply]
> 4. My board also has a PCF8574 io expander and a ADS1115 device. The Tasmota drivers append stuff after the device number, so I get "ADS1115-48" and "PCF8574-1" in the JSON (returned via MQTT). The dash ("-") messes with the templating. 

See https://tasmota.github.io/docs/Commands/#setoption64

**Comment by barbudor** [comment]
@tjaart-github I'm interested in the PCF8583 too as a counter for battery powered rain gauge but haven't purchased the chip yet.
I can give a help if you want

**Reply by HouseOnTheHill** [reply]
Cool. What did you have in mind - a driver?

**Reply by barbudor** [reply]
Yes, 
I don't yet have the chip for testing but I can setup a frame.
Have you already made a test sketch in Arduino IDE for example
You can contact me on Discord through the Tasmota Discord server

**Reply by HouseOnTheHill** [reply]
Thank you. I will do this and then get in touch.
Thanks again.

**Comment by borsa93** [comment]
Hi, I'm still having some trouble trying to use red data (in script) with ESP01s and ads1115. (I'm not able to set it in the right way). can you please help me? Thanks

**Reply by gemu2015** [reply]
not enough info.
please post your script and logging

**Reply by borsa93** [reply]
thanks for the answer, I try to attach the script (tryed to use the example from the site), but I'm quite sure that the issue is related to the address (anyway if I try in the console I2cscan I see that is found the device in 0x48)

script:
>D
val=0
found=0
>B
; check device on I2C bus Nr.2
found=ia2(0x48)

>S
; if found read registers
if found>0 {
val=ir(0x5a)
print %val%
}

console:
20:54:03.074 script compressed to 113 bytes = 82 %
20:54:03.077 Script: nv=2, tv=0, vns=10, ram=48
20:54:03.080 var not found: ia2(0x48)

Thanks for the support

**Reply by barbudor** [reply]
I looks like yo used `ia2()` whereas gemu examples uses `ia()`
ESP8266 only has 1 I2C bus, so no `ia2()` available

**Reply by borsa93** [reply]
thanks for the help, I tryed but it's still not working (same log)

**Reply by barbudor** [reply]
Are you sure you have built correctly Tasmota with scripts and i2c
Please share your user_config_override.h

**Reply by gemu2015** [reply]
ia2 selects the 2. i2c bus on ESP32, on ESP8266 ia and ia2 both select bus number 1.
but since you get "var not found" this feature is not compiled in your firmware !
you must use the last DEV version and enable USE_SCRIPT_I2C

**Reply by borsa93** [reply]
I correct the post. Now it is working, thanks everyone. the thing is now.. what is the right setting to read the analogue ingress values? (i think that should be with ir(reg) ) thanks

S
; if found read registers
if found>0 {
val=ir(0x5a)
print %val%
}

**Reply by gemu2015** [reply]
the ADS1115 has only 4 registers. (0 ...3, see data sheet) so 0x5a makes no sense
1. you have to setup the conversion mode, and multiplexer, reg 1
2. you may then read the word values from reg 0

remark:
 currently you can only write 1 byte at a time. since this chip has only word registers and you have to write words, this may fail.
 i will implement word writes in a future version.
why don't you use tasmotas build in ADS driver ?

**Reply by gemu2015** [reply]
i updated the i2c support to write up to 3 bytes.
iw2(reg val) writes a 16 bit word to device which you need for ADS1115

[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/6445395/xdrv_10_scripter.ino.zip)

**Reply by borsa93** [reply]
I really thank you for the support. I flashed the firmware with the xdrv_10_scripter.ino that you sent to me, but unfortunately I'm having the issue that I'm not able to use the values read from the ADS1115, or maybe I'm not able to set it up correctly (for my ignorance), so I kindly ask if you can help me wrtiting the part of the code that I'm missing. Please note that my goal is to get values read from ADS1115 in a formula and publish the result via MQTT (what I'm missing is how to get the values).

Thanks again

**Reply by gemu2015** [reply]
why don't you use the built in ADS driver ?
the i2c script driver is meant for usage with currently unsupported chips only.
what do you want to achieve ?
for implementing a driver yourself you have to read the chips data sheet and understand how to code.

**Reply by borsa93** [reply]
have you got some docs on how to do it?
I would like to convert the value read from the ADS1115 in something else with formulas (for example from volts to lux, I've already a working formula but I don't know how to use it in tasmota). Thanks

**Reply by gemu2015** [reply]
this is extremely easy, just install the standard ADS driver from Tasmota and here is an example script how to convert values

\>D
raw1=0
raw2=0
raw3=0
raw4=0
lux=0

\>T
; get raw values from converter
raw1=ADS1115#A0
raw2=ADS1115#A1
raw3=ADS1115#A2
raw4=ADS1115#A3

; convert to desired unit
lux=raw1*1050/733

; send to MQTT
\>J
,"LUX":%lux%

; show on WEBUI
\>W
Lux{m}%lux% lux

**Reply by borsa93** [reply]
ooh, that's really easy!! I'm really sorry for the loose of time, but unfortunately I was not able to get it.
Thank you very much for the support, have a good Sunday
